managed-pointer/cpp-20/D0567R1.md

```diff
@@ -5,222 +5,384 @@
+As data movement is tightly coupled with executtion, it is nessesary for a unified interface for data movement to be integrated with the unified interface for execution. This means such a unified interface must integrate seemlessly with the current design philosophy of executors, in that it must be compatible for the range of control structures which the executors abstraction supports and must support te range of platforms which executors can target.
 
-At any one time the managed memory allocation can exist in the memory regions of the host CPU and any number of **execution contexts**. However, the managed memory allocation may only be accessible in one of these memory regions at any given time. This memory region is said to be the accessible memory region. If the accessible memory region is of the host CPU, the host CPU is said to be accessible and if the accessible memory region is of an **execution context** that **execution context** is said to be the accessible **execution context**.
+At the same time a unified interface for data movement must exist at a high level of abstraction and must avoid introducing too many requirements on executor implementations, in order to avoid introducing complex requirements for heterogeneous or distributed memory architectures into C\+\+.
 
-For the host CPU to be accessible or for an **execution context** to be the accessible **execution context**, if it is not already, a synchronisation operation is required. A synchronisation operation is an implementation defined asynchronous operation which moves the data from the currently accessible memory region to another memory region. From the point at which a synchronisation operation is triggered the currently accessible memory region is no longer accessible. Once the synchronisation point is complete the memory region the data is being moved to is now the accessible memory region.
+### 2.5. Must be interoperable between executor implementations
 
-Synchronisation operations are course grained synchronisation in that they synchronise the entire managed memory allocation of a ``managed_ptr``.
+The unified interface for execution will have may different implementation which target different platforms and therefore a unified interfacec for data movement would too require this range of implementations.
 
-*Note: some systems are capable of providing finer-grained synchronisation via atomic operations, however, this is generally only when there is shared virtual memory addressing or shared physical memory. This feature is discussed in more detail in the future work section.*
+It's important therefore that a unified interface for data movement be interoperable between different executor implementations, it must be possible to synchronise data between the memory region of one executor implementation and another.
```

Comment:
You raise a good question here, this is something which perhaps needs to be better described. So this proposal will require each implementation of an executor to provide implementations of the synchronization functions `put` and `get` for any host based device which they support synchronization from (this will generally be a host CPU but does not have to be). These functions will perform an implementation defined set of commands to make a `managed_ptr` available on the memory region of an executor or on the memory region of the host respectively. When synchronizing directly between two executors of different implementations these will have to be synchronized via a host by calling `get` on the source executor and `put` on the destination executor. This is the default behavior of the `sync` function, however this can be optimized by providing an implementation defined channel which is specialized for synchronizing between those executors.